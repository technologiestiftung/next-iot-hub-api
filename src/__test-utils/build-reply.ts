import { FastifyReply } from "fastify";

export const buildReply: (
  overrides?: Record<string, unknown>
) => FastifyReply = (overrides) => {
  const fastifyReply: unknown = {
    code: 200,
    status: jest.fn(() => fastifyReply),
    send: jest.fn(() => fastifyReply),
    ...overrides,
  };
  return fastifyReply as FastifyReply;
};
