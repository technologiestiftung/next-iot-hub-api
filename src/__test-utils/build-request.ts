import { FastifyRequest } from "fastify";

export const buildRequest: (
  overrides?: Record<string, unknown>
) => FastifyRequest = (overrides) => {
  const req: unknown = { ...overrides };
  return req as FastifyRequest;
};
