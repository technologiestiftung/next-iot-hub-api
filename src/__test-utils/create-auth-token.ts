import faker from "faker";
import { FastifyInstance } from "fastify";
import {
  CreateTokenFullResponse,
  authtokenEndpoint,
  supabaseAnonKey,
} from "./index";

export const createAuthToken: (opts: {
  server: FastifyInstance;
  userToken: string;
  getFullResponse?: boolean;
}) => Promise<string | CreateTokenFullResponse> = async ({
  server,
  userToken,
  getFullResponse = false,
}) => {
  const responseToken = await server.inject({
    method: "POST",
    url: authtokenEndpoint,
    headers: {
      authorization: `Bearer ${userToken}`,
      apikey: supabaseAnonKey,
    },
    payload: {
      description: faker.random.words(5),
    },
  });

  const resBody = JSON.parse(responseToken.body);
  return getFullResponse ? resBody.data : resBody.data.token;
};
