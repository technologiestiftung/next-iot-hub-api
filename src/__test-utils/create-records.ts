// Copyright (c) 2021 Technologiestiftung Berlin & Fabian Mor√≥n Zirfas
//
// This software is released under the MIT License.
// https://opensource.org/licenses/MIT

import { supabase } from "./index";
import { definitions } from "../common/supabase";
// const binomial = require("@stdlib/random/base/binomial");

type Record = Omit<definitions["records"], "measurements"> & {
  measurements: number[];
};

function recordDataBuilder({
  sensor_id,
  num,
}: {
  sensor_id: number;
  num: number;
}): Omit<Record, "id">[] {
  const data: Omit<Record, "id">[] = [];
  for (let i = 0; i < num; i++) {
    data.push({
      recorded_at: new Date().toISOString(),
      // TODO: replace with random values
      measurements: [Math.random() * 10],
      sensor_id,
    });
  }
  return data;
}
export async function createRecords({
  sensor_id,
  numberOfRecords,
}: {
  sensor_id: number;
  numberOfRecords: number;
}): Promise<Record[]> {
  const { data, error } = await supabase
    .from("records")
    .insert(recordDataBuilder({ sensor_id, num: numberOfRecords }));
  if (error) {
    throw error;
  }
  if (!data) {
    throw new Error("No data returned");
  }
  return data;
}
