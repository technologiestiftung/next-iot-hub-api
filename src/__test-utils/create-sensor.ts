import faker from "faker";
import { definitions } from "../common/supabase";
import { supabase } from "./index";

export const createSensor: (options: {
  user_id: string;
  name?: string;
  external_id?: string;
}) => Promise<definitions["sensors"]> = async ({
  user_id,
  name,
  external_id,
}) => {
  const { data: sensors, error: dError } = await supabase
    .from<definitions["sensors"]>("sensors")
    .insert([
      {
        name: name ? name : faker.random.words(2),
        user_id,
        external_id,
        category_id: 1,
        latitude: parseFloat(faker.address.latitude()),
        longitude: parseFloat(faker.address.longitude()),
        altitude: faker.datatype.number({ min: 0, max: 100, precision: 0.1 }),
      },
    ]);
  if (!sensors) {
    throw dError;
  }
  return sensors[0];
};
