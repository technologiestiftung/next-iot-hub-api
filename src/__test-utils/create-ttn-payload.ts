import { TTNPostBody } from "../integrations/ttn";

export function createTTNPayload(
  overrides?: TTNPostBody | Record<string, unknown>
): TTNPostBody {
  return {
    simulated: true,
    end_device_ids: {
      application_ids: {
        application_id: "foo",
      },

      device_id: "123",
    },
    received_at: new Date().toISOString(),
    uplink_message: {
      decoded_payload: { measurements: [1, 2, 3], bytes: [1, 2, 3] },
      locations: {
        user: {
          latitude: 13,
          longitude: 52,
          altitude: 23,
          source: "SOURCE_REGISTRY",
        },
      },
    },
    ...overrides,
  };
}
