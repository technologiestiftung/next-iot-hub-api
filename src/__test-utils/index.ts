import faker from "faker";
import config from "config";
import { createClient } from "@supabase/supabase-js";
import { FastifyInstance, FastifyReply, FastifyRequest } from "fastify";
import fetch from "node-fetch";
import { definitions } from "../common/supabase";
import { TTNPostBody } from "../integrations/ttn";
export type Sensor = definitions["sensors"];

export const jwtSecret =
  process.env.JWT_SECRET ||
  "super-secret-jwt-token-with-at-least-32-characters-long";
export const supabaseAnonKey = process.env.SUPABASE_ANON_KEY || "123";
export const apiVersion = config.get("apiVersion");
export const supabaseServiceRoleKey =
  process.env.SUPABASE_SERVICE_ROLE_KEY || "123";
export const supabaseUrl = "http://localhost:8000";
export const authtokenEndpoint = `/api/v${apiVersion}/authtokens`;
const supabase = createClient(supabaseUrl, supabaseServiceRoleKey);

export const buildServerOpts = {
  jwtSecret,
  supabaseUrl,
  supabaseServiceRoleKey,
  logger: false,
  issuer: "tsb",
};
export interface CreateTokenFullResponse {
  token: string;
  nice_id: number;
}
export interface ApiAuthResponsePayload {
  access_token: string;
  token_type: string;
  expires_in: number;
  refresh_token: string;
  user: User;
}

export interface User {
  id: string;
  aud: string;
  role: string;
  email: string;
  confirmed_at: Date;
  last_sign_in_at: Date;
  app_metadata: AppMetadata;
  user_metadata: null;
  created_at: Date;
  updated_at: Date;
}

export interface JWTPayload {
  sub: string;
  projectId: number;
  description: string;
  jti: string;
  iss: string;
  iat: number;
}

export interface AppMetadata {
  provider: string;
}

export const buildReply: (
  overrides?: Record<string, unknown>
) => FastifyReply = (overrides) => {
  const fastifyReply: unknown = {
    code: 200,
    status: jest.fn(() => fastifyReply),
    send: jest.fn(() => fastifyReply),
    ...overrides,
  };
  return fastifyReply as FastifyReply;
};

export const buildRequest: (
  overrides?: Record<string, unknown>
) => FastifyRequest = (overrides) => {
  const req: unknown = { ...overrides };
  return req as FastifyRequest;
};

interface SignupLoginResponse {
  token: string;
  id: string;
}

export const signupUser: () => Promise<{
  id: string;
  token: string;
}> = async () => {
  const { id, token } = await signup({
    anonKey: supabaseAnonKey,
    email: `${faker.random.word()}+${faker.internet.email()}`,
    password: faker.internet.password(),
    url: new URL(`${supabaseUrl}/auth/v1/signup`),
  });
  return { id, token };
};
export const signup: (options: {
  email: string;
  password: string;
  url: URL;
  anonKey: string;
}) => Promise<SignupLoginResponse> = async ({
  email,
  password,
  url,
  anonKey,
}) => {
  const headers: HeadersInit = {
    "Content-Type": "application/json",
    apikey: anonKey,
  };
  const body = JSON.stringify({ email, password });
  const response = await fetch(url.href, {
    method: "POST",
    headers,
    body,
  });
  if (!response.ok) {
    throw new Error(await response.text());
  }
  if (response.status === 200) {
    const json = (await response.json()) as ApiAuthResponsePayload;
    return { token: json.access_token, id: json.user.id };
  } else {
    throw new Error(await response.json());
  }
};
export const login: (options: {
  email: string;
  password: string;
  url: URL;
  anonKey: string;
}) => Promise<SignupLoginResponse> = async ({
  email,
  password,
  url,
  anonKey,
}) => {
  const headers: HeadersInit = {
    "Content-Type": "application/json",
    apikey: anonKey,
  };
  const body = JSON.stringify({ email, password });

  const response = await fetch(url.href, {
    method: "POST",
    headers,
    body,
  });
  if (!response.ok) {
    throw new Error(await response.text());
  }
  if (response.status === 200) {
    const json = (await response.json()) as ApiAuthResponsePayload;
    return { token: json.access_token, id: json.user.id };
  } else {
    throw new Error(await response.json());
  }
};

export const logout: (options: {
  userToken: string;
  url: URL;
  anonKey: string;
}) => Promise<boolean> = async ({ url, userToken, anonKey }) => {
  const headers: HeadersInit = {
    "Content-Type": "application/json",
    apikey: anonKey,
    Authorization: `Bearer ${userToken}`,
  };

  const response = await fetch(url.href, {
    method: "POST",
    headers,
  });
  if (!response.ok) {
    throw new Error(await response.text());
  }
  if (response.status === 204) {
    return true;
  } else {
    throw new Error(await response.json());
  }
};

export const deleteUser: (userToken: string) => Promise<boolean> = async (
  userToken
) => {
  const headers: HeadersInit = {
    "Content-Type": "application/json",
    apikey: supabaseAnonKey,
    Authorization: `Bearer ${userToken}`,
  };

  const url = new URL(`${supabaseUrl}/rest/v1/rpc/delete_user`);
  const response = await fetch(url.href, {
    method: "POST",
    headers,
  });
  if (!response.ok) {
    throw new Error(await response.text());
  }
  if (response.status !== 200) {
    console.error(await response.json());
    throw new Error("Could not delete user");
  }
  return true;
};

/**
 *
 * @deprecated
 */
export const createProject: (options: {
  name?: string;
  userId: string;
  categoryId?: number;
}) => Promise<void> = async () => {
  return;
};

export const createSensor: (options: {
  user_id: string;
  name?: string;
  external_id?: string;
}) => Promise<definitions["sensors"]> = async ({
  user_id,
  name,
  external_id,
}) => {
  const { data: sensors, error: dError } = await supabase
    .from<definitions["sensors"]>("sensors")
    .insert([
      {
        name: name ? name : faker.random.word(),
        user_id,
        external_id,
        category_id: 1,
        latitude: parseFloat(faker.address.latitude()),
        longitude: parseFloat(faker.address.longitude()),
        altitude: faker.datatype.number({ min: 0, max: 100, precision: 0.1 }),
      },
    ]);
  if (!sensors) {
    throw dError;
  }
  return sensors[0];
};

export const createAuthToken: (opts: {
  server: FastifyInstance;
  userToken: string;
  getFullResponse?: boolean;
}) => Promise<string | CreateTokenFullResponse> = async ({
  server,
  userToken,
  getFullResponse = false,
}) => {
  const responseToken = await server.inject({
    method: "POST",
    url: authtokenEndpoint,
    headers: {
      authorization: `Bearer ${userToken}`,
      apikey: supabaseAnonKey,
    },
    payload: {
      description: faker.random.words(5),
    },
  });

  const resBody = JSON.parse(responseToken.body);
  return getFullResponse ? resBody.data : resBody.data.token;
};

export const createTTNPayload: (
  overrides?: TTNPostBody | Record<string, unknown>
) => TTNPostBody = (overrides) => {
  return {
    simulated: true,
    end_device_ids: {
      application_ids: {
        application_id: "foo",
      },

      device_id: "123",
    },
    received_at: new Date().toISOString(),
    uplink_message: {
      decoded_payload: { measurements: [1, 2, 3], bytes: [1, 2, 3] },
      locations: {
        user: {
          latitude: 13,
          longitude: 52,
          altitude: 23,
          source: "SOURCE_REGISTRY",
        },
      },
    },
    ...overrides,
  };
};
