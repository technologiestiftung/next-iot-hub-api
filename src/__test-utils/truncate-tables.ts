import { Pool } from "pg";
import { databaseUrl } from "../lib/env";

export const pool = new Pool({
  connectionString: databaseUrl,
});
export async function truncateTables(): Promise<void> {
  try {
    const client = await pool.connect();
    await client.query("TRUNCATE auth.users restart identity cascade");
    await client.query(
      "TRUNCATE public.user_profiles restart identity cascade"
    );
    client.release();
  } catch (error) {
    console.error("DB error while truncating", error);
    throw error;
  }
}
