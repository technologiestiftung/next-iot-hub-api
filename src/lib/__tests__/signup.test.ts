/* eslint-disable jest/no-hooks */
// Copyright (c) 2021 Technologiestiftung Berlin & Fabian MorÃ³n Zirfas
//
// This software is released under the MIT License.
// https://opensource.org/licenses/MIT

import faker from "faker";
import {
  apiVersion,
  buildServerOpts,
  checkInbox,
  deleteUser,
  purgeInbox,
  signupUser,
  truncateTables,
} from "../../__test-utils";
import { closePool } from "../../__test-utils/truncate-tables";
import buildServer from "../server";

const signupUrl = `/api/v${apiVersion}/signup`;
describe("signup POST tests", () => {
  beforeEach(async () => {
    await truncateTables();
  });
  // eslint-disable-next-line jest/no-hooks
  afterAll(async () => {
    await truncateTables();
    await closePool();
  });
  test("should be rejected due to missing body", async () => {
    const server = buildServer(buildServerOpts);
    const response = await server.inject({
      method: "POST",
      url: signupUrl,
    });
    expect(response.statusCode).toBe(400);
  });
  test("should be rejected due to missing body property email", async () => {
    const server = buildServer(buildServerOpts);
    const response = await server.inject({
      method: "POST",
      url: signupUrl,
      payload: {
        user: "baz",
      },
    });
    expect(response.statusCode).toBe(400);
  });
  test("should be rejected due to missing body property user", async () => {
    const server = buildServer(buildServerOpts);
    const response = await server.inject({
      method: "POST",
      url: signupUrl,
      payload: {
        email: faker.internet.email(),
      },
    });
    expect(response.statusCode).toBe(400);
  });
  test("should be rejected due to username already taken", async () => {
    const server = buildServer(buildServerOpts);
    const { token, userProfile } = await signupUser("ff6347");
    // if (!userProfile) throw new Error("Could not create userProfile");

    const response = await server.inject({
      method: "POST",
      url: signupUrl,
      payload: {
        email: faker.internet.email(),
        name: userProfile?.name,
      },
    });

    expect(response.statusCode).toBe(409);
    await deleteUser(token);
  });

  test("should be rejected due to email already taken", async () => {
    const server = buildServer(buildServerOpts);
    const name = "ff6347";
    const email = "me@email.com";
    const { token } = await signupUser(name, email);

    const response = await server.inject({
      method: "POST",
      url: signupUrl,
      payload: {
        email,
        name: faker.random.word(), //?
      },
    });
    expect(response.statusCode).toBe(409);
    await deleteUser(token);
  });

  test("should create a new user and user profile with a username", async () => {
    const server = buildServer(buildServerOpts);
    const name = "ff6347";
    const email = "me@email.com";
    await purgeInbox(email.split("@")[0]);
    const response = await server.inject({
      method: "POST",
      url: signupUrl,
      payload: {
        email,
        name,
      },
    });
    const messages = await checkInbox("me");
    expect(messages).toHaveLength(1);
    expect(messages).toMatchSnapshot([
      {
        date: expect.any(String),
        from: "<info@stadtpuls.com>",
        id: expect.any(String),
        mailbox: "me",
        "posix-millis": expect.any(Number),
        seen: false,
        size: expect.any(Number),
        subject: "Your Magic Link",
        to: ["<me@email.com>"],
      },
    ]);
    expect(response.statusCode).toBe(204);
  });
});
