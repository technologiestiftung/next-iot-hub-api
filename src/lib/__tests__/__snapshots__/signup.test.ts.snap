// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`signup POST tests should create a new user and user profile with a username 1`] = `
Array [
  Object {
    "date": Any<String>,
    "from": "<info@stadtpuls.com>",
    "id": Any<String>,
    "mailbox": "me",
    "posix-millis": Any<Number>,
    "seen": false,
    "size": 539,
    "subject": "Your Magic Link",
    "to": Array [
      "<me@email.com>",
    ],
  },
]
`;
